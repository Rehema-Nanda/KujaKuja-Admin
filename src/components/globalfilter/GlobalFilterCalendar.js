import React from "react";

import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import moment from "moment";
import "./GlobalFilter.css";
import GlobalFilterCalendarContainer from "./GlobalFilterCalendarContainer";
import CalendarConfig from "./CalendarConfig";

export default class GlobalFilterCalendar extends GlobalFilterCalendarContainer {
    constructor(props) {
        super(props);

        this.handleDateChangeStart = this.handleDateChangeStart.bind(this);
        this.handleDateChangeEnd = this.handleDateChangeEnd.bind(this);
        this.datePickerInputFocusHandler = this.datePickerInputFocusHandler.bind(this);
        this.calendarContainer = this.calendarContainer.bind(this);
        this.state = {
            selectedDateRangeType: this.getDateRangeType(),
            isMobileView: false,
        };
    }

    handleDateChangeStart(startDate) {
        const { dateEnd } = this.props;

        this.handleDateRangeSelect(startDate, dateEnd, CalendarConfig.IS_START_DATE);
    }

    handleDateChangeEnd(endDate) {
        const { dateStart } = this.props;

        this.handleDateRangeSelect(dateStart, endDate, CalendarConfig.IS_END_DATE);
    }

    datePickerInputFocusHandler() {
        // immediately blur on focus to prevent the virtual keyboard from displaying on mobile devices
        // TODO: test for side effects, for example if the 'withPortal' option is true (see setBlur, cancelFocusInput, deferFocusInput & handleBlur functions at
        //       https://github.com/Hacker0x01/react-datepicker/blob/master/src/index.jsx)
        if (this.refs) {
            if (this.refs[CalendarConfig.START_DATE_PICKER_REF]) {
                this.refs[CalendarConfig.START_DATE_PICKER_REF].setBlur();
            }
            if (this.refs[CalendarConfig.END_DATE_PICKER_REF]) {
                this.refs[CalendarConfig.END_DATE_PICKER_REF].setBlur();
            }
        }
    }

    render() {
        const { dateEnd, dateStart, withPortal } = this.props;
        const { selectedDateRangeType } = this.state;

        return (

            <div className="date-pickers-holder">

                <button
                    type="button"
                    onClick={() => {
                        this.selectNewDateRange(-1);
                    }}
                    className="left-right-button left-button"
                >
                    <div className="left-right-arrow left-arrow" />
                </button>

                <DatePicker
                    className="date-picker-startdate"
                    withPortal={withPortal}
                    selected={dateStart}
                    selectsStart
                    ref={CalendarConfig.START_DATE_PICKER_REF}
                    monthsShown={2}
                    useWeekdaysShort={true}
                    onFocus={this.datePickerInputFocusHandler}
                    startDate={dateStart}
                    endDate={dateEnd}
                    maxDate={moment.utc().startOf("day")}
                    calendarContainer={this.calendarContainer}
                    popperPlacement="bottom-start"
                    popperModifiers={{
                        preventOverflow: {
                            enabled: true,
                            escapeWithReference: false,
                            boundariesElement: "viewport",
                        },
                    }}
                    onSelect={this.handleDateChangeStart}
                />

                <span className={`dash ${selectedDateRangeType === CalendarConfig.DAILY ? "hidden" : ""}`}> - </span>

                <DatePicker
                    className={`date-picker-enddate ${selectedDateRangeType === CalendarConfig.DAILY ? "hide-date" : ""}`}
                    withPortal={withPortal}
                    selected={dateEnd}
                    selectsEnd
                    ref={CalendarConfig.END_DATE_PICKER_REF}
                    monthsShown={2}
                    useWeekdaysShort={true}
                    onFocus={this.datePickerInputFocusHandler}
                    startDate={dateStart}
                    endDate={dateEnd}
                    maxDate={moment.utc().startOf("day")}
                    calendarContainer={this.calendarContainer}
                    popperPlacement="bottom-start"
                    popperModifiers={{
                        preventOverflow: {
                            enabled: true,
                            escapeWithReference: false,
                            boundariesElement: "viewport",
                        },
                    }}
                    onSelect={this.handleDateChangeEnd}
                />

                <button
                    type="button"
                    onClick={() => {
                        this.selectNewDateRange(1);
                    }}
                    className="left-right-button right-button"
                >
                    <div className="left-right-arrow right-arrow" />
                </button>

            </div>

        );
    }
}
