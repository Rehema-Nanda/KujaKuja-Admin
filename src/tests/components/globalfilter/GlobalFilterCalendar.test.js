import React from "react";
import { shallow } from "enzyme";
import moment from "moment";
import GlobalFilterCalendar from "../../../components/globalfilter/GlobalFilterCalendar";

describe("<GlobalFilterCalendar />", () => {
    const testProps = {
        dateEnd: moment.utc("2018-12-12"),
        setDateStartEnd: jest.fn(),
        dateStart: moment.utc("2018-01-01")
    };
    const wrapper = shallow(<GlobalFilterCalendar {...testProps} />);
    it("renders correctly", () => {
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it("calls handleDateChangeStart function when date in datepicker is selected", () => {
        const handleDateChangeStart = jest.spyOn(wrapper.instance(), "handleDateChangeStart");
        const handleDateChangeEnd = jest.spyOn(wrapper.instance(), "handleDateChangeEnd");
        wrapper.instance().forceUpdate();
        const startDatePicker = wrapper.find("DatePicker").first();
        const endDatePicker = wrapper.find("DatePicker").at(1);
        startDatePicker.simulate("select");
        endDatePicker.simulate("select");
        expect(handleDateChangeStart).toHaveBeenCalled();
        expect(handleDateChangeEnd).toHaveBeenCalled();
    });

    it("calls selectNewDateRange when button is clicked", () => {
        const selectNewDateRangeFunc = jest.spyOn(wrapper.instance(), "selectNewDateRange");
        wrapper.instance().forceUpdate();
        const button = wrapper.find("button").first();
        const secondButton = wrapper.find("button").at(1);
        button.simulate("click");
        secondButton.simulate("click");
        expect(selectNewDateRangeFunc).toHaveBeenCalled();
    });

    it("calls datePickerInputFocusHandler when the focus changes", () => {
        const datePickerInputFocusHandlerFunc = jest.spyOn(wrapper.instance(), "datePickerInputFocusHandler");
        wrapper.instance().forceUpdate();
        const startDatePicker = wrapper.find("DatePicker").first();
        startDatePicker.simulate("focus");
        expect(datePickerInputFocusHandlerFunc).toHaveBeenCalled();
    });
});
