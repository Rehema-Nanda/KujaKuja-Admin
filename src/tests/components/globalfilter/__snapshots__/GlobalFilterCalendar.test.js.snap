// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalFilterCalendar /> renders correctly 1`] = `
"<div className=\\"date-pickers-holder\\">
  <button type=\\"button\\" onClick={[Function: onClick]} className=\\"left-right-button left-button\\">
    <div className=\\"left-right-arrow left-arrow\\" />
  </button>
  <DatePicker className=\\"date-picker-startdate\\" withPortal={false} selected={{...}} selectsStart={true} monthsShown={2} useWeekdaysShort={true} onFocus={[Function: bound datePickerInputFocusHandler]} startDate={{...}} endDate={{...}} maxDate={{...}} calendarContainer={[Function: bound calendarContainer]} popperPlacement=\\"bottom-start\\" popperModifiers={{...}} onSelect={[Function: bound handleDateChangeStart]} allowSameDay={false} dateFormat=\\"L\\" dateFormatCalendar=\\"MMMM YYYY\\" onChange={[Function: onChange]} disabled={false} disabledKeyboardNavigation={false} dropdownMode=\\"scroll\\" onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyDown]} onInputClick={[Function: onInputClick]} onClickOutside={[Function: onClickOutside$$1]} onMonthChange={[Function: onMonthChange]} preventOpenOnFocus={false} onYearChange={[Function: onYearChange]} onInputError={[Function: onInputError]} readOnly={false} shouldCloseOnSelect={true} showTimeSelect={false} timeIntervals={30} timeCaption=\\"Time\\" previousMonthButtonLabel=\\"Previous Month\\" nextMonthButtonLabel=\\"Next month\\" renderDayContents={[Function: renderDayContents]} />
  <span className=\\"dash \\">
     - 
  </span>
  <DatePicker className=\\"date-picker-enddate \\" withPortal={false} selected={{...}} selectsEnd={true} monthsShown={2} useWeekdaysShort={true} onFocus={[Function: bound datePickerInputFocusHandler]} startDate={{...}} endDate={{...}} maxDate={{...}} calendarContainer={[Function: bound calendarContainer]} popperPlacement=\\"bottom-start\\" popperModifiers={{...}} onSelect={[Function: bound handleDateChangeEnd]} allowSameDay={false} dateFormat=\\"L\\" dateFormatCalendar=\\"MMMM YYYY\\" onChange={[Function: onChange]} disabled={false} disabledKeyboardNavigation={false} dropdownMode=\\"scroll\\" onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyDown]} onInputClick={[Function: onInputClick]} onClickOutside={[Function: onClickOutside$$1]} onMonthChange={[Function: onMonthChange]} preventOpenOnFocus={false} onYearChange={[Function: onYearChange]} onInputError={[Function: onInputError]} readOnly={false} shouldCloseOnSelect={true} showTimeSelect={false} timeIntervals={30} timeCaption=\\"Time\\" previousMonthButtonLabel=\\"Previous Month\\" nextMonthButtonLabel=\\"Next month\\" renderDayContents={[Function: renderDayContents]} />
  <button type=\\"button\\" onClick={[Function: onClick]} className=\\"left-right-button right-button\\">
    <div className=\\"left-right-arrow right-arrow\\" />
  </button>
</div>"
`;
